Always use FMT_STRING internally where possible [Issue #2002]

yeswalrus opened the pull request:
Update a few internal calls to fmt::format to use FMT_STRING, improving compatibility with the FMT_ENFORCE_COMPILE_STRING compile option.

I agree that my contributions are licensed under the {fmt} license, and agree to future changes to the licensing.

vitaut commented on include/fmt/color.h:
```diff
@@ -578,7 +578,7 @@ inline std::basic_string<Char> vformat(
 template <typename S, typename... Args, typename Char = char_t<S>>
 inline std::basic_string<Char> format(const text_style& ts, const S& format_str,
-                                      const Args&... args) {
+                                      Args&&... args) {
```
This looks unrelated, please revert.

yeswalrus replied:
Done, though this was done since I noticed it would bring the function more into alignment with the other examples of and documentation for functions using `make_args_checked`

vitaut commented on include/fmt/chrono.h:
```diff
@@ -764,13 +764,20 @@ inline std::chrono::duration<Rep, std::milli> get_milliseconds(
+  using context = FMT_BUFFER_CONTEXT(type_identity_t<Char>);
+
   const Char pr_f[] = {'{', ':', '.', '{', '}', 'f', '}', 0};
-  if (precision >= 0) return format_to(out, pr_f, val, precision);
+  if (precision >= 0) {
+    return vformat_to(out, to_string_view(pr_f),
+                      make_format_args<context>(val, precision));
+  }
```
Please make `pr_f``static constexpr` and keep `format_to` (don't change it to `vformat_to`). `static constexpr` strings should work with `FMT_STRING`:

```c++
static constexpr const Char pr_f[] = {'{', ':', '.', '{', '}', 'f', '}', 0};
  if (precision >= 0) return format_to(out, FMT_STRING(pr_f), val, precision);
```

Same below.

yeswalrus replied:
ah, good to know! Won't that break on platforms where FMT_CONSTEXPR_DECL evaluates to noop though?

vitaut replied:
On that platform `FMT_STRING` should also evaluate to noop (at least in theory).

yeswalrus replied:
Nm, tested on godbolt and realized that's a perfectly valid c++11 feature

yeswalrus replied:
This apparently breaks on windows, even after correcting it to use SFINAE to avoid compiling the `{:.{}f}` format string with an integral type. I'm adding a test for using FMT_STRING with a static FMT_CONSTEXPR_DECL, and I'll see if I can get a workaround

vitaut reviewed (changes requested):
Thanks for the PR. Looks good in general but please address the inline comment.

yeswalrus commented:
I know you asked me not to modifiy tests, but I've added a demonstration of a failure to format-test.cc starting at line 1819, and filed a bug report against MSVC++ here: https://developercommunity.visualstudio.com/content/problem/1256077/internal-compiler-failure-with-macro-defining-lamb.html
I've modified chrono.h to use enable_if in a sadly combinatoric way to avoid having to use FMT_STRING on a constexpr char array and to avoid compiling a compile-time checked format string with a bad type.

yeswalrus commented:
I've simplified by directly calling the underlying function, detail::compile_string_to_view, to work around the MSVC crash. It simplifies things significantly, so LMK which version you'd prefer.

vitaut reviewed (changes requested):
Mostly looks good but please address inline comments.

yeswalrus commented on test/format-test.cc:
```diff
@@ -1807,11 +1807,27 @@ fmt::string_view to_string_view(string_like) { return "foo"; }
   EXPECT_EQ(L"42", fmt::format(FMT_STRING(L"{}"), 42));
   EXPECT_EQ("foo", fmt::format(FMT_STRING("{}"), string_like()));
+
+  (void)static_with_null;
+  (void)static_with_null_wide;
+  (void)static_no_null;
+  (void)static_no_null_wide;
+#if !defined(_MSC_VER)
+  EXPECT_EQ("42", fmt::format(FMT_STRING(static_with_null), 42));
+  EXPECT_EQ(L"42", fmt::format(FMT_STRING(static_with_null_wide), 42));
+  EXPECT_EQ("42", fmt::format(FMT_STRING(static_no_null), 42));
+  EXPECT_EQ(L"42", fmt::format(FMT_STRING(static_no_null_wide), 42));
+#endif
```
Not quite - you test with constexpr locals, which are a c++ 17 feature, but static constexpr char arrays are valid all the way back to c++11 and this exposes a bug in MSVC.

vitaut commented:
Thank you!
